File: partial_function_application.swift
#!/usr/bin/env swift
// http://rosettacode.org/wiki/Partial_function_application

func newMap(f: (Int) -> (Int), s: [Int], v: [Int] = [], i: Int = 0) -> [Int] {
  var newS = v
  if i == count(s) {
    return v
  } else {
    newS.append(f(s[i]))
    return newMap(f, s, v: newS, i: i+1)
  }
}

let f1 = { (n: Int) -> Int in return n * 2 }
let f2 = { (n: Int) -> Int in return n * n }
let fs = { (f: (Int) -> (Int), s: Int) -> Int in return f(s) }
let fsf1 = { (s: [Int]) -> [Int] in return newMap(f1,s) }
let fsf2 = { (s: [Int]) -> [Int] in return newMap(f2,s) }

for list in [[0,1,2,3],[2,4,6,8]] {
  println(fsf1(list))
  println(fsf2(list))
}

File: partial_function_application.rb
#!/usr/bin/env ruby
# http://rosettacode.org/wiki/Partial_function_application

new_map = -> f,s,i=0 { i == s.count ? (return s) : (s[i] = f[s[i]]; return new_map[f,s,i+1]) }

f1 = -> n { n * 2 }
f2 = -> n { n * n }
fs = -> f,s { f[s] }
fsf1 = -> s { new_map[f1,s] }
fsf2 = -> s { new_map[f2,s] }

[[0,1,2,3],[2,4,6,8]].each { |e| p fsf1[e]; p fsf2[e] }
